Fix service health check node port leaks

girishkalele opened the pull request:
Fixed #31757

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31990)

<!-- Reviewable:end -->

Commit:
Fix service health check node port leaks

smarterclayton commented on pkg/registry/service/rest.go:
```diff
@@ -157,7 +157,7 @@ func (rs *REST) Create(ctx api.Context, obj runtime.Object) (runtime.Object, err
 		}
 	}
 
-	if featuregate.DefaultFeatureGate.ExternalTrafficLocalOnly() && shouldCheckOrAssignHealthCheckNodePort(service) {
```
Why remove the gate check?

girishkalele replied:
The feature gate check already exists in that function being called (`shouldCheckOrAssignHealthCheckNodePort), so removed the redundant check here.

https://github.com/girishkalele/kubernetes/blob/master/pkg/registry/service/rest.go#L458

smarterclayton commented on pkg/registry/service/rest.go:
```diff
@@ -265,6 +275,97 @@ func (*REST) NewList() runtime.Object {
+			if err != nil {
+				errmsg := fmt.Sprintf("Failed to allocate requested HealthCheck nodePort %v:%v",
+					requestedHealthCheckNodePort, err)
+				el := field.ErrorList{field.Invalid(field.NewPath("ObjectMeta", "Annotations"),
```
Fieldpaths should be for dominant external version name, "metadata", "annotations", etc.

girishkalele replied:
Ok, will change the fieldpath to "annotations".

smarterclayton replied:
Maybe github is broken but i don't see the change - should be `field.NewPath("metadata", "annotations")` in all these checks

girishkalele commented:
@smarterclayton Updated. PTAL.

smarterclayton commented on pkg/registry/service/rest.go:
```diff
@@ -234,6 +234,16 @@ func (rs *REST) Delete(ctx api.Context, id string) (runtime.Object, error) {
+			err := rs.serviceNodePorts.Release(int(nodePort))
+			if err != nil {
+				// these should be caught by an eventual reconciliation / restart
+				glog.Errorf("Error releasing service health check %s node port %d: %v", service.Name, nodePort, err)
```
use `utilruntime.HandleError` anytime you eat an error (that allows us to plug in error shipping plugins and track them externally).  The default behavior is glog.Errorf so is exactly equal

smarterclayton commented:
One comment

smarterclayton commented on pkg/registry/service/rest.go:
```diff
@@ -342,6 +443,11 @@ func (rs *REST) Update(ctx api.Context, name string, objInfo rest.UpdatedObjectI
 		service.Status.LoadBalancer = api.LoadBalancerStatus{}
 	}
 
+	success, err := rs.healthCheckNodePortUpdate(oldService, service)
```
Should this be feature gated?

girishkalele replied:
Good point - yes, I need to add feature gates to the new calls in Delete and Update.

girishkalele replied:
Actually, Delete already has the feature gate. healthCheckNodePortUpdate() internally has the feature gate check, but it is not complete, will fix that up.

girishkalele commented:
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/registry/service/rest.go, line 446 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31990#-KQg-Icrg7f5JUDHTBrp-r2-446:-KR5TXtjyP29qPNpeaxI:b-lf0b2o) ([raw file](https://github.com/kubernetes/kubernetes/blob/4e5d08a3a11ce9d626c03927fd6940ab3b869971/pkg/registry/service/rest.go#L446)):*
<details><summary><i>Previously, smarterclayton (Clayton Coleman) wrote…</i></summary>
> Should this be feature gated?
</details>
Good point - yes, I need to add feature gates to the new calls in Delete and Update.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31990)*
<!-- Sent from Reviewable.io -->

girishkalele commented on pkg/registry/service/rest.go:
```diff
@@ -265,6 +276,97 @@ func (*REST) NewList() runtime.Object {
+	//   Reject deletion of the "service.alpha.kubernetes.io/healthcheck-nodeport" annotation
+	//   Reject changing the value of the healthCheckNodePort annotation
+	//
+	oldServiceHasHealthCheckNodePort := shouldCheckOrAssignHealthCheckNodePort(oldService)
```
Added missing feature gate check here.

girishkalele commented on pkg/registry/service/rest.go:
```diff
@@ -265,6 +276,97 @@ func (*REST) NewList() runtime.Object {
+	case oldServiceHasHealthCheckNodePort && assignHealthCheckNodePort:
+		if _, ok := service.Annotations[apiservice.AnnotationHealthCheckNodePort]; !ok {
+			glog.Warningf("Attempt to delete health check node port annotation DENIED")
+			el := field.ErrorList{field.Invalid(field.NewPath("metadata", "annotations"),
```
Github not broken, it was me, I forgot to add that to the commit. Fixed now.

girishkalele commented on pkg/registry/service/rest.go:
```diff
@@ -234,6 +235,16 @@ func (rs *REST) Delete(ctx api.Context, id string) (runtime.Object, error) {
+			err := rs.serviceNodePorts.Release(int(nodePort))
+			if err != nil {
+				// these should be caught by an eventual reconciliation / restart
+				utilruntime.HandleError(fmt.Errorf("Error releasing service health check %s node port %d: %v", service.Name, nodePort, err))
```
Changed to utilruntime.HandleError here.

thockin commented:
ping @smarterclayton 1.4 blocker

thockin commented:
Actually, I take that back, this should not block 1.4.0, the feature is alpha anyway, still..

bprashanth commented on pkg/registry/service/rest.go:
```diff
@@ -157,7 +158,7 @@ func (rs *REST) Create(ctx api.Context, obj runtime.Object) (runtime.Object, err
 	}
 
-	if featuregate.DefaultFeatureGate.ExternalTrafficLocalOnly() && shouldCheckOrAssignHealthCheckNodePort(service) {
+	if shouldCheckOrAssignHealthCheckNodePort(service) {
```
you have a comment below that I don't understand "Insert the newly allocated health check port as an annotation (plan of record for Alpha)", can you clarify it.

bprashanth commented on pkg/registry/service/rest.go:
```diff
@@ -157,7 +158,7 @@ func (rs *REST) Create(ctx api.Context, obj runtime.Object) (runtime.Object, err
+	if shouldCheckOrAssignHealthCheckNodePort(service) {
 		var healthCheckNodePort int
 		var err error
 		if l, ok := service.Annotations[apiservice.AnnotationHealthCheckNodePort]; ok {
```
Doesn't `GetServiceHealthCheckNodePort` do this for you? maybe you just return an err and int32 as the distinction between "no node port specified" and "node port specified but couldn't parse".  What do we think should happen if user sets it to 0? currently we clobber it. Maybe we should throw an error instead? 

Leave it to you to do what you think is cleaner, just bringing it up because I observed duplicated code.

bprashanth commented on pkg/registry/service/rest.go:
```diff
@@ -265,6 +276,97 @@ func (*REST) NewList() runtime.Object {
+	oldServiceHasHealthCheckNodePort := shouldCheckOrAssignHealthCheckNodePort(oldService)
+	oldHealthCheckNodePort := apiservice.GetServiceHealthCheckNodePort(oldService)
+
+	assignHealthCheckNodePort := shouldCheckOrAssignHealthCheckNodePort(service)
```
the name of this function is confusing and makes me think it's doing 2 different things. Can you clarify the name or add a comment above it. It returns a bool, so the shouldCheck part doesn't make sense to me.

bprashanth commented on pkg/registry/service/rest.go:
```diff
@@ -265,6 +276,97 @@ func (*REST) NewList() runtime.Object {
+					apiservice.AnnotationHealthCheckNodePort, errmsg)}
+				return false, errors.NewInvalid(api.Kind("Service"), service.Name, el)
+			}
+			glog.Infof("Reserved user requested nodePort: %d", requestedHealthCheckNodePort)
```
Not sure if anything else uses Infof.V(2) in apiserver logs. If you open it up all you'll see is Http requests and latency traces. Suggest converting to V(4) everywhere in its currently V(2) in this pr.

bprashanth commented on pkg/registry/service/rest.go:
```diff
@@ -265,6 +276,97 @@ func (*REST) NewList() runtime.Object {
+				apiservice.AnnotationHealthCheckNodePort, "Cannot change healthcheck nodePort during update")}
+			return false, errors.NewInvalid(api.Kind("Service"), service.Name, el)
+		}
+	}
```
it's useful to add a comment here saying this is the "normal" update path, i.e the else clause is what happens if i modify labels on a Service using esipp.

bprashanth commented on pkg/registry/service/rest.go:
```diff
@@ -265,6 +276,97 @@ func (*REST) NewList() runtime.Object {
+			el := field.ErrorList{field.Invalid(field.NewPath("metadata", "annotations"),
+				apiservice.AnnotationHealthCheckNodePort, "Cannot insert healthcheck nodePort annotation")}
+			return false, errors.NewInvalid(api.Kind("Service"), service.Name, el)
+		}
```
worth adding a comment that this is what happens in the normal Service update case when the feature is gated (so every service on aws, azure etc).

bprashanth commented:
I'd really like a unittest for the big switch case you have, otherwise i see mostly nits. Would be good to get this for 1.4 because otherwise toggling the annotation will leak nodeports. I'll take a follow up for the unittests if you prefer.

girishkalele commented:
Review status: 0 of 1 files reviewed at latest revision, 12 unresolved discussions.

---

*[pkg/registry/service/rest.go, line 161 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31990#-KQg-Icrg7f5JUDHTBrp-r3-161:-KRFezezevUAj6QatETv:b-u0n6xl) ([raw file](https://github.com/kubernetes/kubernetes/blob/02754547fbbc9110358763aa38a9d3770af691f4/pkg/registry/service/rest.go#L161)):*
<details><summary><i>Previously, bprashanth (Prashanth B) wrote…</i></summary>
> you have a comment below that I don't understand "Insert the newly allocated health check port as an annotation (plan of record for Alpha)", can you clarify it.
</details>
I had there to remind me it needs to turn into an API field instead of annotation for GA.

---

*[pkg/registry/service/rest.go, line 164 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31990#-KQg-Icrg7f5JUDHTBrp-r3-164:-KRFfrL6BJ-9T9QhOkFn:bwj18md) ([raw file](https://github.com/kubernetes/kubernetes/blob/02754547fbbc9110358763aa38a9d3770af691f4/pkg/registry/service/rest.go#L164)):*
<details><summary><i>Previously, bprashanth (Prashanth B) wrote…</i></summary>
> Doesn't `GetServiceHealthCheckNodePort` do this for you? maybe you just return an err and int32 as the distinction between "no node port specified" and "node port specified but couldn't parse".  What do we think should happen if user sets it to 0? currently we clobber it. Maybe we should throw an error instead? 
> 
> 
> 
> Leave it to you to do what you think is cleaner, just bringing it up because I observed duplicated code. 
</details>
Yes, I thought I couldn't use the existing `GetServiceHealthCheckNodePort` because this code runs in the rest request context and returns NewInternalErrors(). I wouldn't want to change the `GetServiceHealthCheckNodePort` function at this stage, too many callers around. Maybe a TODO, but this is going to change from annotation to a service object field for GA (I think).

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31990)*
<!-- Sent from Reviewable.io -->

girishkalele commented:
Review status: 0 of 1 files reviewed at latest revision, 11 unresolved discussions.

---

*[pkg/registry/service/rest.go, line 243 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31990#-KQg-Icrg7f5JUDHTBrp-r2-243:-KRFgWIQjWyn_YbQGAES:bb6un1j) ([raw file](https://github.com/kubernetes/kubernetes/blob/4e5d08a3a11ce9d626c03927fd6940ab3b869971/pkg/registry/service/rest.go#L243)):*
<details><summary><i>Previously, smarterclayton (Clayton Coleman) wrote…</i></summary>
> use `utilruntime.HandleError` anytime you eat an error (that allows us to plug in error shipping plugins and track them externally).  The default behavior is glog.Errorf so is exactly equal
</details>
Done

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31990)*
<!-- Sent from Reviewable.io -->

girishkalele commented:
Review status: 0 of 1 files reviewed at latest revision, 10 unresolved discussions.

---

*[pkg/registry/service/rest.go, line 303 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31990#-KQg-Icrg7f5JUDHTBrp-r3-303:-KRFglCB7I_iQPr4CDL3:b-sw6jd1) ([raw file](https://github.com/kubernetes/kubernetes/blob/02754547fbbc9110358763aa38a9d3770af691f4/pkg/registry/service/rest.go#L303)):*
<details><summary><i>Previously, bprashanth (Prashanth B) wrote…</i></summary>
> the name of this function is confusing and makes me think it's doing 2 different things. Can you clarify the name or add a comment above it. It returns a bool, so the shouldCheck part doesn't make sense to me. 
</details>
Renamed to `serviceNeedsHealthCheckNodePort`.

---

*[pkg/registry/service/rest.go, line 319 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31990#-KQg-Icrg7f5JUDHTBrp-r3-319:-KRFgyA1AVFcNMavQ_m9:bxj60ro) ([raw file](https://github.com/kubernetes/kubernetes/blob/02754547fbbc9110358763aa38a9d3770af691f4/pkg/registry/service/rest.go#L319)):*
<details><summary><i>Previously, bprashanth (Prashanth B) wrote…</i></summary>
> Not sure if anything else uses Infof.V(2) in apiserver logs. If you open it up all you'll see is Http requests and latency traces. Suggest converting to V(4) everywhere in its currently V(2) in this pr. 
</details>
Good point - changed all glog.Infof to glog.V(4).Infof.

---

*[pkg/registry/service/rest.go, line 351 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31990#-KQg-Icrg7f5JUDHTBrp-r3-351:-KRFhYfhyUEBkdzmkOt9:b-8b11j9) ([raw file](https://github.com/kubernetes/kubernetes/blob/02754547fbbc9110358763aa38a9d3770af691f4/pkg/registry/service/rest.go#L351)):*
<details><summary><i>Previously, bprashanth (Prashanth B) wrote…</i></summary>
> worth adding a comment that this is what happens in the normal Service update case when the feature is gated (so every service on aws, azure etc). 
</details>
Added

---

*[pkg/registry/service/rest.go, line 366 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31990#-KQg-Icrg7f5JUDHTBrp-r3-366:-KRFi2jd1552OlHE0W-q:b-8b11j9) ([raw file](https://github.com/kubernetes/kubernetes/blob/02754547fbbc9110358763aa38a9d3770af691f4/pkg/registry/service/rest.go#L366)):*
<details><summary><i>Previously, bprashanth (Prashanth B) wrote…</i></summary>
> it's useful to add a comment here saying this is the "normal" update path, i.e the else clause is what happens if i modify labels on a Service using esipp. 
</details>
Added

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31990)*
<!-- Sent from Reviewable.io -->

bprashanth commented:
LGTM because I want to try and get this into the next beta so we can have test time (though it might be tight). Please send unittests pr as follow up. Looks like you addressed all @smarterclayton 's comments as well.

bprashanth commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.kubernetes.io/lgtm.png" height="20" width="61"/>

---







*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31990#-:-KRFqsjweWKkBBSY0oEK:bnfp4nl)*
<!-- Sent from Reviewable.io -->

girishkalele commented:
@bprashanth Yes, I will take the unittests as a follow-up PR to this one.
